utils/self-extracting-executable/decompressor.cpp

```diff
@@ -371,38 +347,42 @@ int main(int argc, char* argv[])
     if (0 != close(input_fd))
         perror(nullptr);
 
-    /// According to documentation `mv` will rename file if it
-    /// doesn't move to other directory.
-    /// Sometimes `rename` doesn't exist by default and
-    /// `rename.ul` is set instead. It will lead to errors
-    /// that can be easily avoided with help of `mv`
-
-    if (!have_compressed_analoge)
+    if (unlink(argv[0]))
     {
-        printf("No target executable - decompression only was performed.\n");
-        /// remove file
-        execlp("rm", "rm", argv[0], NULL);
         perror(nullptr);
         return 1;
     }
+
+    if (!have_compressed_analoge)
+        printf("No target executable - decompression only was performed.\n");
     else
     {
-        /// move decompressed file instead of this binary and apply command
-        char bash[] = "sh";
-        char executable[] = "-c";
-
-        /// length of forwarded args
-        size_t length = 0;
-        for (int i = 1; i < argc; ++i)
-            length += strlen(argv[i]);
-
-        /// mv filename.decompressed filename && filename agrs...
-        char command[3 + strlen(argv[0]) + 14 + strlen(argv[0]) + 4 + strlen(argv[0]) + length + argc];
-        fillCommand(command, argc, argv, length);
-
-        /// replace file and call executable
-        char * newargv[] = { bash, executable, command, nullptr };
-        execvp(bash, newargv);
+        const char * const decompressed_name_fmt = "%s.decompressed.%s";
```

Comment:
(Surely a matter of taste but the code would overall be more readable and equally portable if it was completely in (standard) C++. Neither bloat nor performance will be a problem. But it is likely too late for that now ^^).

Reply:
It's not too late, you can rewrite everything in C++.
The only concern for the decompressor is its size, but up to a few MB is normal.

Reply:
:) I will think about it... but I'm not very enthusiastic regarding this at all :)
from other side pureness of C is already broken here :(